### PR TITLE
feat(linter): add noAutofocus rule for HTML

### PR DESCRIPTION
**Pull request by tt-a1i**
## Summary

Port the `noAutofocus` a11y rule to HTML. This rule enforces that the `autofocus` attribute is not used on elements, as it can cause usability issues for sighted and non-sighted users.

The rule allows `autofocus` in the following contexts:
- Inside `<dialog>` elements
- Inside elements with the `popover` attribute

Part of #8155

## Test plan

- Added `invalid.html` test with 5 cases that should trigger the rule
- Added `valid.html` test with dialog and popover exception cases
- All tests pass with `cargo test -p biome_html_analyze no_autofocus`

## AI Assistance Disclosure

I used Codex to review the changes, sanity-check the implementation against existing patterns, and help spot potential edge cases.

### CODE REVIEW

File: crates/biome_html_analyze/src/lint/a11y/no_autofocus.rs [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+fn is_autofocus_attribute(node: &HtmlAttribute) -> bool {
+    node.name().is_ok_and(|name| {
+        name.value_token()
+            .is_ok_and(|value_token| value_token.text_trimmed().eq_ignore_ascii_case("autofocus"))
+    })
+}
+
+/// Check if the element is inside an allowed context (dialog or popover)
+///
+/// Note: We skip the first element (the one containing the autofocus attribute)
```
Comment (ematipico) [inline crates/biome_html_analyze/src/lint/a11y/no_autofocus.rs]:
```suggestion
/// Note: We skip the first [HtmlElement] (the one containing the autofocus attribute)
```

File: crates/biome_html_analyze/src/lint/a11y/no_autofocus.rs [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+/// because we only want to check if it's *inside* a dialog/popover, not if
+/// it *is* the dialog/popover itself.
+fn is_inside_allowed_context(attr: &HtmlAttribute) -> bool {
```
Comment (ematipico) [inline crates/biome_html_analyze/src/lint/a11y/no_autofocus.rs]: Return `Option<bool>`, so that you can convert all `let Ok()` in `.ok()?`

File: crates/biome_html_analyze/src/lint/a11y/no_autofocus.rs [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+                    && token.text_trimmed().eq_ignore_ascii_case("dialog")
+                {
+                    return true;
+                }
+                // Check if element has popover attribute
+                if opening.find_attribute_by_name("popover").is_some() {
+                    return true;
+                }
+            }
+        }
```
Comment (ematipico) [inline crates/biome_html_analyze/src/lint/a11y/no_autofocus.rs]: We're repeating these checks, which means this code isn't well-maintainable. Instead, let's do something more generic.

We already have `AnyHtmlTagElement` from `biome_html_syntax`, let's use that so we can extract the `name` of the element and then do one single check

**Comment by tt-a1i** [comment]
> Looks good. The code can be improved

hi, thank you for your review, i fix these issue already

File: crates/biome_html_analyze/src/lint/a11y/no_autofocus.rs [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+
+/// Extract an [AnyHtmlTagElement] from an ancestor node
+fn get_tag_element(node: &biome_html_syntax::HtmlSyntaxNode) -> Option<AnyHtmlTagElement> {
+    HtmlElement::cast_ref(node)
+        .and_then(|e| e.opening_element().ok())
+        .map(AnyHtmlTagElement::from)
+        .or_else(|| {
+            HtmlSelfClosingElement::cast_ref(node).map(|e| AnyHtmlTagElement::from(e.clone()))
+        })
+}
```
Comment (ematipico) [inline crates/biome_html_analyze/src/lint/a11y/no_autofocus.rs:156]: I don't understand this function. You can do `AnyHtmlTagElement::cast_ref(node)` and it will work. Technically, this function isn't needed, and you can use the `cast_ref` inside the loop.

**Comment by tt-a1i** [comment]
Thanks! I did try `AnyHtmlTagElement::cast_ref(node)` first, but it won’t match what we get from `ancestors()`.

For normal tags like `<div popover>…</div>`, the ancestor node is `HtmlElement`, and the attributes live on its `opening_element` (`HtmlOpeningElement`). `AnyHtmlTagElement` only covers `HtmlOpeningElement` / `HtmlSelfClosingElement`, so casting a `HtmlElement` directly always returns `None`.

That’s why I keep `get_tag_element()`: it’s just the small adapter that turns `HtmlElement -> opening_element` (and also handles self-closing tags) so the rest of the check can be written once.

**Comment by ematipico** [comment]
I understand now, thank you
